linalg/hypre.hpp

```diff
@@ -956,6 +1003,106 @@ class HypreEuclid : public HypreSolver
+/// \note There are several variants available depending on the ilu_type chosen:
+/// ilu_type = 0: ILU(k) locally and block Jacobi globally
+/// ilu_type = 1: ILUT locally and block Jacobi globally
+/// ilu_type = 10: ILU(k) locally with inexact GMRES solution of Schur system
+/// ilu_type = 11: ILUT locally with inexact GMRES solution of Schur system
+/// ilu_type = 20: ILU(k) locally with Newton-Schultz-Hotelling (NSH) on Schur
+/// ilu_type = 21: ILUT locally with Newton-Schultz-Hotelling (NSH) on Schur
+/// ilu_type = 30: ILU(k) locally with restricted additive Schwarz globally
+/// ilu_type = 31: ILUT locally with restricted additive Schwarz globally
+/// ilu_type = 40: ILU(k) locally with GMRES with ddPQ ordering for Schur
+/// ilu_type = 41: ILUT locally with GMRES with ddPQ ordering for Schur
+class HypreILU : public HypreSolver
+{
+private:
+   HYPRE_Solver ilu_precond;
+
+   /// The type of incomplete LU used locally and globally (see class doc)
+   HYPRE_Int ilu_type;
+   /// Maximum iterations; 1 iter for preconditioning
+   HYPRE_Int max_iter;
+   /// The tolerance when used as a smoother; set to 0.0 for preconditioner
+   HYPRE_Real tol;
+   /// Fill level for ILU(k)
+   HYPRE_Int lev_fill;
+   /// Maximum non-zeros per row (for ILUT only)
+   HYPRE_Int nz_max;
+   /// Drop tolerance for ILUT
+   HYPRE_Real drop_thres;
+   /// Drop tol in Newton–Schulz–Hotelling iteration
+   HYPRE_Real nsh_thres;
+   /// Maximum number of iter to solve Schur system
+   HYPRE_Int schur_max_iter;
+   /// Local reordering scheme; 0 = no reordering, 1 = reverse Cuthill-McKee
+   HYPRE_Int reorder_type;
+   /// Information print level; 0 = none, 1 = setup, 2 = solve, 3 = setup+solve
+   HYPRE_Int print_level;
+
+   /// Set the ILU default options
+   void SetDefaultOptions();
+
+   /// Set ILU options based on the values saved in data members of this wrapper
+   void SetOptions();
+
+   /// Reset the ILU preconditioner
+   /// \note If ilu_precond is NULL, this method allocates it and sets default
+   /// options. Otherwise the method destroys ilu_precond, allocates a new
+   /// object, and sets its options based on the data members of *this
+   void ResetILUPrecond();
+
+public:
+   /// Constructor; sets the default options
+   HypreILU();
+
+   virtual ~HypreILU();
+
+   /// Set the maximum number of iterations; set max_iter = 1 for preconditioner
+   void SetMaxIter(HYPRE_Int max_iter);
+
+   /// Set the tolerance when used as a smoother
+   void SetTol(HYPRE_Real tol);
+
+   /// Set the fill level for ILU(k) variant
+   void SetLevelOfFill(HYPRE_Int lev_fill);
+      
+   /// Set the maximum number of non-zers per row (needed by ILUT variant)
+   void SetMaxNnzPerRow(HYPRE_Int nz_max);
+
+   /// Set the drop tolerance threshold used by the ILUT variant
+   void SetDropThreshold(HYPRE_Real drop_thres);
+
+   /// Set drop tol used in the Newton-Schulz-Hotelling solution of Schur system
+   void SetNSHDropThreshold(HYPRE_Real nsh_thres);
```

Comment:
I agree here. It is basically unnecessary code duplication without a benefit on the interface, since every function is just a 1:1 map to the internal hypre options.

Reply:
I think this reflects my ignorance that `HYPRE_Solver` cast was even available for this purpose.  I will remove the less common options, but I will also add a comment at the beginning of the class that illustrates how to access the Hypre options directly, ok?  The fact that I could not set the fill level directly in the Hypre Euclid wrapper was very confusing and frustrating (which is why I added all these set functions), so a comment could help future users.

Reply:
Of course you can add any valuable information to the comments!